### PR TITLE
4435 secure sock thread

### PR DESCRIPTION
**Pull request by aleks-f**

### CODE REVIEW

File: NetSSL_OpenSSL/src/SecureSocketImpl.cpp [hunk 1]
```diff
@@ -300,6 +306,9 @@ void SecureSocketImpl::shutdown()
 			int rc = ::SSL_shutdown(_pSSL);
 #endif
 			if (rc < 0) handleError(rc);
+
+			l.unlock();
```
Comment (matejk) [inline NetSSL_OpenSSL/src/SecureSocketImpl.cpp:310]: Why is mutex unlocked explicitly before end of scope only in this function and not also others?

File: NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h [hunk 2]
```diff
@@ -284,16 +284,21 @@ class NetSSL_API SecureSocketImpl
...
 
-	SSL* _pSSL;
+	std::atomic<SSL*> _pSSL;
 	Poco::AutoPtr<SocketImpl> _pSocket;
 	Context::Ptr _pContext;
 	bool _needHandshake;
 	std::string _peerHostName;
 	Session::Ptr _pSession;
 	bool _bidirectShutdown = true;
+	mutable MutexT _mutex;
```
Comment (matejk) [inline NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h:301]: Is `_mutex` intended to protect only _pSSL or also other related members (`_pContext`, `_pSession`)?

**Review by matejk — APPROVED** [review]
Changes seem OK to me. I added few comments where some things were not clear to me.

File: NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h (hunk 2 above)
Comment (aleks-f) [reply NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h:301]: mutex protects all openssl calls; `_pSSL` pointer was made atomic, smart pointers have atomic refcounting

**Pull request merged**